bindings/rust/src/lazy_reader.rs

```diff
@@ -0,0 +1,488 @@
+        }
+    }
+
+    pub fn add_offset(&mut self, offset: usize) {
+        self.offset = self.offset.checked_add(offset).unwrap();
+    }
+
+    pub fn sub_size(&mut self, shrink_size: usize) {
+        self.size = self.size.checked_sub(shrink_size).unwrap();
+    }
+
+    pub fn validate(&self) -> Result<(), Error> {
+        if let Some(size) = self.offset.checked_add(self.size) {
+            if size > self.data_source.borrow().total_size {
+                Err(Error::TotalSize(
+                    "validate: size > cur.source.total_size".into(),
+                ))
+            } else {
+                Ok(())
+            }
+        } else {
+            Err(Error::Overflow("validate".into()))
+        }
+    }
+
+    pub fn unpack_number(&self) -> Result<usize, Error> {
+        let mut src = [0u8; 4];
+        let size = read_at(self, &mut src[..]).unwrap();
+        if size != 4 {
+            Err(Error::FieldCount("unpack_number".into()))
+        } else {
+            let res = u32::from_le_bytes(src);
+            Ok(res as usize)
+        }
+    }
+
+    pub fn verify_fixed_size(&self, total_size: usize) -> Result<(), Error> {
+        if self.size == total_size {
+            Ok(())
+        } else {
+            Err(Error::TotalSize(format!(
+                "self.size({}) == total_size({})",
+                self.size, total_size
+            )))
+        }
+    }
+
+    pub fn fixvec_verify(&self, item_size: usize) -> Result<(), Error> {
+        if self.size < NUM_T_SIZE {
+            return Err(Error::FieldCount(format!(
+                "fixvec_verify, self.size({}) < NUM_T_SIZE",
+                self.size
+            )));
+        }
+        let item_count = self.unpack_number()?;
+        if item_count == 0 {
+            if self.size == NUM_T_SIZE {
+                return Ok(());
+            } else {
+                return Err(Error::Header(format!(
+                    "self.size({}) == NUM_T_SIZE",
+                    self.size
+                )));
+            }
+        }
+
+        let total_size = calculate_offset(item_size, item_count, NUM_T_SIZE);
+        if self.size == total_size {
+            Ok(())
+        } else {
+            Err(Error::TotalSize(format!(
+                "self.size({}) == total_size({})",
+                self.size, total_size
+            )))
+        }
+    }
+
+    pub fn option_is_none(&self) -> bool {
+        self.size == 0
+    }
+    pub fn fixvec_length(&self) -> Result<usize, Error> {
+        self.unpack_number()
+    }
+
+    pub fn dynvec_length(&self) -> Result<usize, Error> {
+        if self.size == NUM_T_SIZE {
+            Ok(0)
+        } else {
+            let mut cur2 = self.clone();
+            cur2.add_offset(NUM_T_SIZE);
+            cur2.sub_size(NUM_T_SIZE);
+            cur2.validate()?;
+            cur2.get_item_count()
+        }
+    }
+
+    pub fn get_item_count(&self) -> Result<usize, Error> {
+        let count = self.unpack_number()? / 4;
```

Comment:
If we define `table Table1 { f1: Bytes, f2: Bytes }`, an valid serialized bytes of a `Table1 { f1: 0x, f2: 0xabcdef }` is
  ```
  # the full size is 23 (0x17) bytes
  17 00 00 00
  # two offsets (8 bytes in total)
  0c 00 00 00, 10 00 00 00,
  # two items (11 bytes in total)
  00 00 00 00
  03 00 00 00, ab cd ef
  ```

but current implementation does not check the validity of the offsets, if we build a malicious slice by offsetting the items by one, the lazy reader still parses it as a valid `Table1`:
  ```
  18 00 00 00
  0d 00 00 00, 11 00 00 00,
  00,
  00 00 00 00
  03 00 00 00, ab cd ef
  ```

I think an additional checking needs to be added to determine if offset mod 4 is 0,